juniper/src/integrations/jiff.rs

```diff
@@ -282,6 +334,51 @@ mod duration {
     }
 }
 
+/// Representation of time zone.
+///
+/// Is a set of rules for determining the civil time, via an offset from UTC, in a particular
+/// geographic region. In many cases, the offset in a particular time zone can vary over the course
+/// of a year through transitions into and out of daylight saving time.
+///
+/// [IANA database][1] compliant.
+///
+/// See also [`jiff::tz::TimeZone`][2] for details.
+///
+/// [1]: http://www.iana.org/time-zones
+/// [2]: https://docs.rs/jiff/latest/jiff/tz/struct.TimeZone.html
+#[cfg(feature = "jiff-tz")]
+#[graphql_scalar(
+    with = time_zone,
+    parse_token(String),
+    specified_by_url = "https://graphql-scalars.dev/docs/scalars/time-zone",
+)]
+pub type TimeZone = jiff::tz::TimeZone;
+
+#[cfg(feature = "jiff-tz")]
+mod time_zone {
+    use super::*;
+
+    pub(super) fn to_output<S>(v: &TimeZone) -> Value<S>
+    where
+        S: ScalarValue,
+    {
+        Value::scalar(
+            v.iana_name()
+                .unwrap_or_else(|| panic!("Failed to format `TimeZone`: no IANA name"))
+                .to_owned(),
```

Comment:
@sgoll non-fallible `to_output` is just the current limitation of the `#[graphql_scalar]`/`#[derive(GraphQLScalar)]` macros expansion and looks like this:
```rust
#[automatically_derived]
impl<__S> ::juniper::GraphQLValue<__S> for jiff::civil::DateTime
where
    __S: ::juniper::ScalarValue,
{
    type Context = ();
    type TypeInfo = ();
    fn type_name<'i>(&self, info: &'i Self::TypeInfo) -> ::core::option::Option<&'i ::core::primitive::str> { <Self as ::juniper::GraphQLType<__S>>::name(info) }
    fn resolve(&self, info: &(), selection: ::core::option::Option<&[::juniper::Selection<'_, __S>]>, executor: &::juniper::Executor<'_, '_, Self::Context, __S>) -> ::juniper::ExecutionResult<__S> {
        ::core::result::Result::Ok(local_date_time::to_output(self))
    }
```
In theory, it shouldn't b a very hard task to add there some `IntoResult` coercion. However, if we do this, the error will be returned to the client. And it's unclear to me whether it's better than just panicking, because failing to serialize a timezone value is not an expected error on the client, and more like something that shouldn't happen at all.

Usually, we solve this via new-type pattern by restricting certain values.

I see 2 options here:
1. Either we create `juniper::integrations::jiff::TimeZone` newtype over `jiff::TimeZone` and restrict constructing its values being non-IANA. This has ergonomics impact as provides a different type.
2. Or we just bother less, leave the `panic!` and mention in docs about the trick above (it's no deal for library users to define their own newtype and `derive(GraphQLScalar)` for it `transparent`ly if they need type safety for that case in their code).

Reply:
Reiterating my ignorance of GraphQL here, using an offset if an IANA name isn't available is fine. It's what RFC 9557 recommends and what Jiff implements itself:

```rust
use jiff::{tz::TimeZone, Zoned};

fn main() -> anyhow::Result<()> {
    let zdt = Zoned::now();

    let tz = TimeZone::get("US/Eastern")?;
    println!("{}", zdt.with_time_zone(tz));

    let tz = jiff::tz::offset(-4).to_time_zone();
    println!("{}", zdt.with_time_zone(tz));

    Ok(())
}
```

Has this output:

```
$ cargo -q r
2024-08-15T20:50:50.12994762-04:00[US/Eastern]
2024-08-15T20:50:50.12994762-04:00[-04:00]
```

Reply:
Can you just use the `Display` impl for `Zoned` for this?

If not, you can use `strftime("%:V")` to get a string that is "IANA time zone name OR offset":

```rust
use jiff::{tz::TimeZone, Zoned};

fn main() -> anyhow::Result<()> {
    let tz = TimeZone::get("US/Eastern")?;
    let zdt = Zoned::now().with_time_zone(tz);
    println!("{}", zdt);
    println!("{}", zdt.strftime("%:V"));

    let tz = jiff::tz::offset(-4).to_time_zone();
    let zdt = Zoned::now().with_time_zone(tz);
    println!("{}", zdt);
    println!("{}", zdt.strftime("%:V"));

    Ok(())
}
```

Has this output:

```
$ cargo -q r
2024-08-15T20:53:52.441756324-04:00[US/Eastern]
US/Eastern
2024-08-15T20:53:52.443774214-04:00[-04:00]
-04:00
```

Note that `%:V` (and `%V`) was introduced in `jiff 0.1.5`.

Reply:
> using an offset if an IANA name isn't available is fine. It's what RFC 9557 recommends

I would be happy to have it, unfortunately `TimeZone` GraphQL scalar implementation is [not capable of that](https://github.com/Urigo/graphql-scalars/blob/v1.23.0/src/scalars/TimeZone.ts). If we provide from the server side wider variety of values than the client could accept, it will break certain invariants and assumptions between them, maybe even silently.

If it only was about input coercion on the server side, this could have been OK, as we do for some `LocalDateTime` scalar implementations, being able to parse non-RFC-compliant `%Y-%m-%d %H:%i:%s` format from clients. However, we're talking here about returning values from server to client, and overcoming client expectations is not a good choice here, imho.

Reply:
@sgoll 

> Where would be a good location to put the mention in the docs regarding `panic!()`?

I think module level docs are fine.

Reply:
@sgoll 

> However, the same effort would be necessary to avoid panicking (if we reverted [ce43466](https://github.com/graphql-rust/juniper/commit/ce434667238f789ca0475222710c17e937427e38)). In both cases, a developer must avoid values that lie outside the strict interpretation of the `TimeZone` scalar,

Sorry, but I don't buy this argument. Because, while the server side is in developer's control here, the client may be not at all. So, having and controlling panics at the server side is much more convenient than having assumptions about clients doing that properly instead. Despite both are not perfect solutions.

> I just learned that there is a separate GraphQL scalar [`UTCOffset`](https://the-guild.dev/graphql/scalars/docs/scalars/utc-offset) which covers exactly those values we are after, after all.

Small remark: it's [named `UtcOffset`](https://github.com/Urigo/graphql-scalars/blob/f3359443959a485900bdaeec1a939ee575aa85a2/src/scalars/UtcOffset.ts#L21) after all. The naming on the site is a little bit fucked up there.

We do use it already for the `time::UtcOffset` type.


> So I propose the following solution:

I'm not against it. Removing subtle footguns is totally OK. I don't use `jiff` in my projects. For our purposes, just pure `time` is more than enough. I just hope those newtypes would not impose unconvenient ergonomics penalties for those who actually use it.

Another approach may be just naming the scalar like `TimeZoneOrUtcOffset` and say in the docs that it's either `TimeZone` or `UtcOffset` value. Actually, it's not another approach. We could provide both newtypes you've described, while making `jiff::tz::TimeZone` still a `TimeZoneOrUtcOffset` scalar.

> In theory, we _could_ also allow building `UTCOffset` scalar values from IANA time zone identifiers but we deliberately decided against it when not implementing [`DateTime`](https://the-guild.dev/graphql/scalars/docs/scalars/date-time) for `Zoned`.

I don't think `DateTime`/`Zoned` is relevant here. We do want to here to narrow the semantics of the `jiff::tz::TimeZone` type to semantics of the `UtcOffset` scalar. The result coercion in such cases is very common in GraphQL land. For example, the `LocalTime` scalar has milliseconds precision, while Rust type, of course, usually have micros and nanos. But we do the coercion to milliseconds when implementing this scalar.

So the situation where we have in the Rust code a `jiff::tz::TimeZone` defined as:
```rust
let tz = TimeZone::get("America/New_York")?;
```
But do want to return it to client as `UtcOffset`, because the client expects it like that, is quite normal and common, as for me, and we should not restrict it, giving the situation that we always may convert IANA to offset.

Reply:
Maybe, we don't even need a second newtype:
- `jiff::tz::TimeZone` implementing `UtcOffset` scalar directly
- `juniper::integrations::jiff::TimeZone` newtype implementing `TimeZone` scalar

Reply:
@sgoll 

> we cannot convert from IANA time zone identifier to UTC offset without also knowing the exact point in time that is relevant.

Ah, yes. My bet. Then let's stick to the solution with 3 scalars.

> The serialization to `UtcOffset` was an addition that became obvious when I worked on the other newtype—which is now not needed anymore.

Ah... ok! Nice! 👍

I'll try to review and merge it today.